### ISSUE TITLE
Trait code failing when using bit vector type

### ISSUE THREAD

**Issue by franck44** [dafny-lang/dafny#549]
The following program 
```
 module M {

    trait T {
         function method bar() : bv8
    }
    class F extends T {   
        function method bar() : bv8 {
            1
        }
    }
 }```

triggers an error on the following command:

dafny /dafnyVerify:0 /compile:0 error.dfy results in:
```
Dafny 2.3.0.10506
[ERROR] FATAL UNHANDLED EXCEPTION: cce+UnreachableException: Exception of type 'cce+UnreachableException' was thrown.
  at Microsoft.Dafny.RefinementTransformer.ResolvedTypesAreTheSame (Microsoft.Dafny.Type prev, Microsoft.Dafny.Type next) [0x003ff] in <b8a04a8c5d6040ac97a2fa1d622f53e6>:0 
  at Microsoft.Dafny.RefinementTransformer.CheckOverride_FunctionParameters (Microsoft.Dafny.Function nw, Microsoft.Dafny.Function f) [0x0004f] in <b8a04a8c5d6040ac97a2fa1d622f53e6>:0 
  at Microsoft.Dafny.Resolver.InheritTraitMembers (Microsoft.Dafny.ClassDecl cl) [0x0059e] in <b8a04a8c5d6040ac97a2fa1d622f53e6>:0 
  at Microsoft.Dafny.Resolver.ResolveTopLevelDecls_Signatures (Microsoft.Dafny.ModuleDefinition def, Microsoft.Dafny.ModuleSignature sig, System.Collections.Generic.List`1[T] declarations, Microsoft.Dafny.Graph`1[Node] datatypeDependencies, Microsoft.Dafny.Graph`1[Node] codatatypeDependencies) [0x003ea] in <b8a04a8c5d6040ac97a2fa1d622f53e6>:0 
  at Microsoft.Dafny.Resolver.ResolveModuleDefinition (Microsoft.Dafny.ModuleDefinition m, Microsoft.Dafny.ModuleSignature sig) [0x00226] in <b8a04a8c5d6040ac97a2fa1d622f53e6>:0 
  at Microsoft.Dafny.Resolver.ResolveProgram (Microsoft.Dafny.Program prog) [0x004e6] in <b8a04a8c5d6040ac97a2fa1d622f53e6>:0 
  at Microsoft.Dafny.Main.Resolve (Microsoft.Dafny.Program program, Microsoft.Dafny.ErrorReporter reporter) [0x00036] in <b8a04a8c5d6040ac97a2fa1d622f53e6>:0 
  at Microsoft.Dafny.Main.ParseCheck (System.Collections.Generic.IList`1[T] files, System.String programName, Microsoft.Dafny.ErrorReporter reporter, Microsoft.Dafny.Program& program) [0x0001e] in <b8a04a8c5d6040ac97a2fa1d622f53e6>:0 
  at Microsoft.Dafny.DafnyDriver.ProcessFiles (System.Collections.Generic.IList`1[T] dafnyFiles, System.Collections.ObjectModel.ReadOnlyCollection`1[T] otherFileNames, Microsoft.Dafny.ErrorReporter reporter, System.Boolean lookForSnapshots, System.String programId) [0x001d8] in <faa1593fe210425eaf9f4a17ce9317fe>:0 
  at Microsoft.Dafny.DafnyDriver.ThreadMain (System.String[] args) [0x0003d] in <faa1593fe210425eaf9f4a17ce9317fe>:0 
  at Microsoft.Dafny.DafnyDriver+<>c__DisplayClass1_0.<Main>b__0 () [0x00001] in <faa1593fe210425eaf9f4a17ce9317fe>:0 
  at System.Threading.ThreadHelper.ThreadStart_Context (System.Object state) [0x00014] in <7d0bc631cadd4a07bf74717847e90cfa>:0 
  at System.Threading.ExecutionContext.RunInternal (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, System.Boolean preserveSyncCtx) [0x00071] in <7d0bc631cadd4a07bf74717847e90cfa>:0 
  at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, System.Boolean preserveSyncCtx) [0x00000] in <7d0bc631cadd4a07bf74717847e90cfa>:0 
  at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state) [0x0002b] in <7d0bc631cadd4a07bf74717847e90cfa>:0 
  at System.Threading.ThreadHelper.ThreadStart () [0x00008] in <7d0bc631cadd4a07bf74717847e90cfa>:0 ```

The error **does not occur** if instead of **bv8** I use int.
The error **does not occur** when the flag **/noResolve** or **/noTypeCheck** is added to the command line arguments.

Are there any restrictions on using bv_x type in traits and classes?

I am using Dafny 2.3.0.10506 on a Mac (using mono).

Thanks for your help.

**Comment by RustanLeino** [comment]
Thanks for the bug report. The bug is a missing case in the handling of method overrides. I have fixed the bug in the midst of some other trait enhancements, and will push that fix when I get the other enhancements into a steady state.

You are using traits and bitvectors correctly. As a temporary workaround, perhaps you can return an `int`, add the postcondition

    ensures 0 <= bar() < 0x100

and change callers to do `bar() as bv8`.

**Comment by franck44** [comment]
Thanks @RustanLeino. I am not actually using this piece of code anymore.
But indeed we used the word around you mentioned but finally opted  for another encoding.
And the work around does not seem to work when operations on bvxx are needed. 
As it is right now, the traits seem to have limited power/usage to coerce some types but fortunately Dafny offers lots of other options and we managed to solve our problem.
I am interested in the new trait enhancements/features when they are released!
